cpp_generator.py

```diff
@@ -0,0 +1,347 @@
+                self.external_includes.add("stdexcept")
+            method_context.write("}\n")
+            method_list.append(method_context.get_content())
+
+        ctor_initializer_list = ",\n".join(ctor_initializer_list)
+
+        # generate class code
+        class_code_context = CodeGeneratorContext()
+        class_code_context.write(f"class {class_name} : public ParentClass {{\n")
+
+        # one indentation is lost after the first line when using replace, that`s why we add an additional one for nested classes
+        if class_name.endswith("Type"):
+            class_code_context.indent()
+        
+        class_code_context.write("public:\n")
+        
+        with class_code_context as public_scope:
+            if name.endswith("Collection"):
+                public_scope.write("%NESTED_CLASSES%\n")
+            public_scope.write(f"{class_name}({ctor_params})")
+            if len(ctor_initializer_list) > 0:
+                with public_scope as ctor_initializer_list_scope:
+                    ctor_initializer_list_scope.write(f":\n{ctor_initializer_list}\n")
+                public_scope.write("{\n}\n\n")
+
+            if len(method_list) > 0:
+                public_scope.write("\n\n".join(method_list))
+                public_scope.write("\n")
+
+            member_list = ";\n".join(member_list)
+            public_scope.write(f"{member_list}" + ";\n")
+
+        class_code_context.write("};\n")
+        return class_code_context.get_content()
+
+    def __gen_collection_types(self, node: VSSNode, path: str) -> str:
+        collection_types = []
+        for child in node.children:
+            if child.type == VSSType.BRANCH:
+                self.includes.add(f"{path}/{child.name}/{child.name}")
+
+                if child.instances:
+                    instances = [(f"{child.name}Collection", None)] + self.__gen_instances(child)
+                    generated_classes = []
+
+                    # create all nested classes for this sub-tree
+                    for i in range(len(instances) - 1):
+                        nested_class = self.__gen_nested_class(child, instances, i)
+                        generated_classes.append(nested_class)
+                        
+                    collection_types.append(generated_classes[0].replace("%NESTED_CLASSES%", "\n\n".join(generated_classes[1:])))
+        
+        return "\n\n".join(collection_types)
+
+    def __gen_model(self, node: VSSNode, path: str, is_root=False):
+        # must be done before generating the imports, since it is adding imports to the list
+        collection_types = self.__gen_collection_types(node, path)
+
+        self.__gen_header(node)
+        self.__gen_imports(node)
+        self.ctx_header.write(self.__generate_opening_namespace_text(self.root_namespace, node))
+        self.ctx_header.write("using ParentClass = Model;\n\n")
```

Comment:
Add a comment here: `Provide an alias for the parent class to avoid name conflicts with members`